### DISCUSSION TITLE
Duplicate files getting generated under bin/debug and Chromium subfolder for CefSharp version 121.3.70

### DISCUSSION THREAD

**Discussion by koder4076** [cefsharp/CefSharp#4785] [General]
We use CefSharp based embedded browser for our solution and keep on upgrading whenever a critical patch is released. Based on a recent upgrade to version 121.3.70, we have found that certain files which were previously present only under the Chromium subfolder is being copied over to the bin/debug folder as well, which results in our installers getting gloated. So essentially, we are having duplicate copies for some files like chrome_100_percent.pak, chrome_200_percent.pak. etc. Refer the image below for details. 

We want to understand the reason for it and come up with a way to remove the duplication of files keeping the functionality intact (maybe using some project level settings or customizing the build process in some way, etc.).

**Chromium Sub-Folder**
![CefSharpFiles_Duplicated](https://github.com/cefsharp/CefSharp/assets/109600420/e8436b0b-d18d-4c23-ab46-66a20b1d6287)

**bin/debug folder**
![CefSharp](https://github.com/cefsharp/CefSharp/assets/109600420/16628c30-42b0-408c-8d3b-db8c64501eda)

**Comment by amaitland** [comment]
> Based on a recent upgrade to version 121.3.70,

What version are you upgrading from?

> only under the Chromium subfolder

How are you copying the files to this subfolder?

### Troubleshooting
Add the following to your `csproj/vbproj` file at the end

```xml
<Target Name="CefSharpAfterBuildDebug" AfterTargets="AfterBuild">
    <CallTarget Targets="CefSharpAfterBuildDiagnostic" />
</Target>
```

Change your `MSBuild project output verbosity` to `Normal`:
```
1. Go to Tools->Options menu in VS
2. Open Projects and Solutions->Build and Run
3. Change the value of the MSBuild project build output verbosity. Pick Normal or higher
```

After you build, check the output window for a block of text that looks like:
```
>Target CefSharpAfterBuildDebug:
>  Target CefSharpAfterBuildDiagnostic:
>    CefSharp After Build Diagnostic
>    CefSharpBuildAction = Content
>    CefSharpPropertiesLoaded = true
>    CefSharpTargetDir =
>    CefSharpTargetDirAnyCpu32 = x86\
>    CefSharpTargetDirAnyCpu64 = x64\
>    RuntimeIdentifier = win7-x64
>    EffectivePlatform =
>    Platform = x64
>    PlatformName = x64
>    Platforms = x86;x64;AnyCPU
>    PlatformTarget = x64
>    CefSharpPlatformTarget = x64
>    PlatformTargetAsMSBuildArchitecture = x64
>    TargetFramework = net472
>    TargetFrameworkVersion = v4.7.2
>    libcef.dll exists = bin.net472\x64\Debug\net472\libcef.dll
>    CefSharpTransformXmlDllPath = C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VisualStudio\v16.0\Web\Microsoft.Web.Publishing.Tasks.dll
>    NuGetProjectStyle = PackageReference
>    NuGetToolVersion = 5.11.1
>    OutputType = WinExe
>    OutDir = bin.net472\x64\Debug\net472\
>    Prefer32Bit = false
```

Copy and paste the output here, making sure to format it for ease of reading.

- https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-text
- https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code

**Reply by koder4076** [reply]
> What version are you upgrading from?

The CEFSharp version is being upgraded from 116.0.230 to 121.3.70

> How are you copying the files to this subfolder?

I have inherited the solution, so I am not very certain of the intricacies of the build process. I assumed that it is a part of the CefSharp build process that it copies the contents to the Chromium subfolder, meaning that I thought that the Chromium subfolder was created by the CefSharp build process itself.

> After you build, check the output window for a block of text that looks like:

The output from the build is as follows
```
CefSharpAfterBuildDiagnostic:
  CefSharp After Build Diagnostic
  CefSharpBuildAction = None
  CefSharpPropertiesLoaded = true
  CefSharpTargetDir = Chromium\
  CefSharpTargetDirAnyCpu32 = Chromium\x86\
  CefSharpTargetDirAnyCpu64 = Chromium\x64\
  CefSharpAnyCpuSupport = true
  RuntimeIdentifier = 
  EffectivePlatform = 
  Platform = x86
  PlatformName = x86
  Platforms = 
  PlatformTarget = x86
  CefSharpPlatformTarget = x86
  PlatformTargetAsMSBuildArchitecture = x86
  TargetFramework = 
  TargetFrameworkVersion = v4.8
  libcef.dll exists = bin\Debug\Chromium\libcef.dll
  CefSharpTransformXmlDllPath = C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Microsoft\VisualStudio\v17.0\Web\Microsoft.Web.Publishing.Tasks.dll
  NuGetProjectStyle = PackageReference
  NuGetToolVersion = 6.9.2
  OutputType = WinExe
  OutDir = bin\Debug\
  Prefer32Bit = true
```

**Comment by amaitland** [comment]
> The CEFSharp version is being upgraded from 116.0.230 to 121.3.70

There was a change to the `Nuget` packages in `121`. For testing purposes please try `120.2.70` and see if the behaviour is the same/different.

**Reply by amaitland** [reply]
I see the problem, the `native` assets are being copied by `msbuild` (where we don't use the `runtimes\win-x86\native` folders prior to `121` for `.Net Framework` nuget packages).

You can manually exclude the `native` assets from the [PackageReference](https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets)

```xml
<!--
For never versions of Visual Studio excluding assets on the top level dependency should work
-->
<ItemGroup>
  <PackageReference Include="CefSharp.WinForms" Version="123.0.60">
    <ExcludeAssets>native</ExcludeAssets>
  </PackageReference>
  <PackageReference Include="CefSharp.OffScreen" Version="123.0.60">
    <ExcludeAssets>native</ExcludeAssets>
  </PackageReference>
</ItemGroup>

<!-- For older versions you might need to exclude native assets on the transitive dependency in this case chromiumembeddedframework.runtime.win-x86  -->
<ItemGroup>
  <PackageReference Include="CefSharp.WinForms" Version="123.0.60"/>
  <PackageReference Include="chromiumembeddedframework.runtime.win-x86" Version="123.0.6">
    <ExcludeAssets>native</ExcludeAssets>
  </PackageReference>
</ItemGroup>

```

I'm taking a guess at the packages you are using from the file structure, you will need to adjust accordingly (including version numbers).

**Reply by koder4076** [reply]
> > The CEFSharp version is being upgraded from 116.0.230 to 121.3.70
> 
> There was a change to the `Nuget` packages in `121`. For testing purposes please try `120.2.70` and see if the behaviour is the same/different.

I did the downgrade to the earlier version and the duplicate files are no longer generated. Only the CefSharp.dll file is generated in both the bin/debug and bin/debug/Chromium folders.



> You can manually exclude the `native` assets from the [PackageReference](https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets)

I tried this (I excluded the native assets in the Package Reference section of the primary project, did not change in other projects which are referenced and also use the cefsharp dlls). But it does not change the result in any way.

I am not sure if I am doing it wrong or I have to do the same change across all projects where the CefSharp dlls are referenced or would it be more convenient to write a post build event attached to the main project and delete the duplicate files.

PFA the screenshot of the csproj file diff.

![CefSharp](https://github.com/cefsharp/CefSharp/assets/109600420/9243631a-3ddc-4559-8acd-12645fac2ff7)

**Reply by amaitland** [reply]
> I am not sure if I am doing it wrong or I have to do the same change across all projects where the CefSharp dlls are referenced

The native assets come from `chromiumembeddedframework.runtime.win-x86` and `chromiumembeddedframework.runtime.win-x64` packages. Depending on your project structure you may need to explicitly exclude the native assets from those packages.

> I tried this (I excluded the native assets in the Package Reference section of the primary project, did not change in other projects which are referenced and also use the cefsharp dlls).

It would entirely depend on how you have included the packages in the project. 

You'll need to read and understand https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets

**Reply by koder4076** [reply]
> The native assets come from `chromiumembeddedframework.runtime.win-x86` and `chromiumembeddedframework.runtime.win-x64` packages. Depending on your project structure you may need to explicitly exclude the native assets from those packages.

Thank you so much for your help. I was finally able to resolve the issue following the link you suggested and the following changes from your suggestion.

> ```
> 
> <ItemGroup>
>   <PackageReference Include="CefSharp.WinForms" Version="123.0.60"/>
>   <PackageReference Include="chromiumembeddedframework.runtime.win-x86" Version="123.0.6">
>     <ExcludeAssets>native</ExcludeAssets>
>   </PackageReference>
> </ItemGroup>
> ```
